### PR TITLE
MM-45120: Track team join times. Add API to retrieve new team members since a given time.

### PR DESCRIPTION
**Pull request by mkraft**
#### Summary

* Adds a new `CreateAt` column to the `TeamMembers` table to track the time users join a team.
* Adds a new btree index to the new `TeamMembers.CreateAt` column.
* Adds a new API endpoint `GET /api/v4/teams/:team_id/top/team_members` to retrieve the team members who have joined the given team (today, or in the past 7 or 28 days).

Example API request:

```shell
curl 'http://localhost:8065/api/v4/teams/okofznhigfna3js4ue9fsrjtma/top/team_members?page=0&per_page=1&time_range=today' \
  -H 'Authorization: Bearer heqd4eawgffsmczgb18m68fkga' \
  -H 'X-Requested-With: XMLHttpRequest'
```

Example response:

```json
{
  "has_next": true,
  "items": [
    {
      "id": "bwmd5gwsnpn39du1chysa4rw6r",
      "username": "albert.torres",
      "first_name": "Albert",
      "last_name": "Torres",
      "nickname": "Lazz",
      "position": "Accountant IV",
      "create_at": 1659016008054
    }
  ],
  "total_count": 6
}
```

Total count query:

```sql
SELECT
    count(*)
FROM
    TeamMembers
    JOIN Users ON Users.id = TeamMembers.userid
    LEFT JOIN Bots ON Bots.userid = Users.id
WHERE
    TeamMembers.createat >= 1658980800000
    AND TeamMembers.deleteat = 0
    AND Users.deleteat = 0
    AND teamid = 'okofznhigfna3js4ue9fsrjtma'
    AND Bots.userid is null;
```

Items query:

```sql
SELECT
    Users.id,
    Users.username,
    Users.firstname,
    Users.lastname,
    Users.position,
    Users.position,
    TeamMembers.createat,
    Users.nickname
FROM
    TeamMembers
    JOIN Users ON Users.id = TeamMembers.userid
    LEFT JOIN Bots ON Bots.userid = Users.id
WHERE
    TeamMembers.createat >= 1658980800000
    AND TeamMembers.deleteat = 0
    AND Users.deleteat = 0
    AND teamid = 'okofznhigfna3js4ue9fsrjtma'
    AND Bots.userid is null
LIMIT
    1 OFFSET 0;
```

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-45120

#### Release Note

```release-note
Started tracking the join time of team members and added a new API endpoint to retrieve information about team members who have joined during a given time.
```

### Related PR

https://github.com/mattermost/mattermost-webapp/pull/10895

### CODE REVIEW

**Comment by BenCookie95** [comment]
@mkraft I think we are just missing nickname so that I can display their preferred name!

File: app/team_test.go [hunk 1]
```diff
@@ -1594,3 +1594,191 @@ func TestInviteGuestsToChannelsGracefully(t *testing.T) {
...
+			_, beforeCount, appErr := th.App.GetNewTeamMembersSince(th.Context, th.BasicTeam.Id, &model.InsightsOpts{StartUnixMilli: 0, Page: 0, PerPage: 1000})
+			require.Nil(t, appErr)
+			appErr = th.App.PermanentDeleteUser(th.Context, anotherUser)
+			require.Nil(t, appErr)
+			_, afterCount, appErr := th.App.GetNewTeamMembersSince(th.Context, th.BasicTeam.Id, &model.InsightsOpts{StartUnixMilli: 0, Page: 0, PerPage: 1000})
+			require.Nil(t, appErr)
+			require.Equal(t, beforeCount-1, afterCount)
+		})
+
+		t.Run("including bots", func(t *testing.T) {
```
Comment (mkraft) [inline app/team_test.go:1672]: TODO: Update this test if we decide that bots should be excluded.

File: app/team_test.go [hunk 2]
```diff
@@ -1594,3 +1594,191 @@ func TestInviteGuestsToChannelsGracefully(t *testing.T) {
...
+			require.Nil(t, appErr)
+			require.Contains(t, nUIDs(beforeMembersList.Items), anotherUser.Id)
+			appErr = th.App.PermanentDeleteUser(th.Context, anotherUser)
+			require.Nil(t, appErr)
+			afterMembersList, _, appErr := th.App.GetNewTeamMembersSince(th.Context, th.BasicTeam.Id, &model.InsightsOpts{StartUnixMilli: 0, Page: 0, PerPage: 1000})
+			require.Nil(t, appErr)
+			require.NotContains(t, nUIDs(afterMembersList.Items), anotherUser.Id)
+		})
+
+		t.Run("including bots", func(t *testing.T) {
```
Comment (mkraft) [inline app/team_test.go:1773]: TODO: Update this test if we decide that bots should be excluded.

File: i18n/en.json [hunk 3]
```diff
@@ -5843,6 +5843,10 @@
     "id": "app.post.get_flagged_posts.app_error",
     "translation": "Unable to get the flagged posts."
   },
+  {
+    "id": "app.post.get_new_team_members_since.app_error",
+    "translation": " "
```
Comment (plant99) [inline i18n/en.json:5848]: Should we fill this translation entry?

**Review by plant99 — APPROVED** [review]
LGTM, just a small request to add translation string.

File: i18n/en.json (hunk 3 above)
Comment (mkraft) [reply i18n/en.json:5848]: It's not being used by the front-end so per can leave it blank to avoid unnecessary work for the translators.

**Comment by BenCookie95** [comment]
@mkraft apologies one thing I missed from the response object for the users is `last_picture_update`. Can we add that in please?

**Comment by BenCookie95** [comment]
@mkraft also, can we use snake case for the user fields?

**Comment by mkraft** [comment]
> @mkraft also, can we use snake case for the user fields?

Done.

**Review by BenCookie95 — APPROVED** [review]
LGTM, just missing `last_picture_update` on the user object if possible

**Review by furqanmlk — APPROVED** [review]
Working as expected and test case has been added. [MM-T4968](https://mattermost.atlassian.net/projects/MM?selectedItem=com.atlassian.plugins.atlassian-connect-plugin:com.kanoah.test-manager__main-project-page#!/project/10302/testCase?folderId=3247244)

**Pull request merged**